python/pfs/utils/optimizeBlackDots.py

```diff
@@ -0,0 +1,535 @@
+
+            # the ones which were seen - add penalty if you should not have seen them
+            quality_measure_observed = []
+            for j in range(len(actual_position_x_observed)):
+                quality_measure_observed.append(self.check(actual_position_x_observed[j],
+                                                           actual_position_y_observed[j], xd, yd, 0.75))
+
+            # the ones which should not be seen - add penalty if you saw them
+            quality_measure_predicted = []
+            for j in range(len(predicted_position_x_not_observed)):
+                # if you were not able to predict position, no penalty
+                if np.isnan(predicted_position_x_not_observed[j]):
+                    val = 0
+                else:
+                    val = self.check(predicted_position_x_not_observed[j],
+                                     predicted_position_y_not_observed[j], xd, yd, 0.75)
+                    val = 1 - val
+                quality_measure_predicted.append(val)
+
+            quality_measure_single_run = np.sum(quality_measure_observed) + np.sum(quality_measure_predicted)
+            list_of_quality_measure_single_run.append(quality_measure_single_run)
+
+        sum_quality_measure_single_run = np.sum(list_of_quality_measure_single_run)
+
+        return sum_quality_measure_single_run
+
+    def total_penalty_for_single_dot(self, i, xd, yd, print_outputs=False):
+        """Calculate a penalty for a single dot
+
+        Wrapper function for `quality_measure_single_spot`, that
+        calculate penatly given index of the spot and the dot position
+
+        Parameters
+        ----------
+        i: `integer`
+            index of the cobra
+        xd: `float`
+            x position of the dot
+        yd: `float`
+            y position of the dot
+        print_outputs: `bool`
+            if True print outputs while computing
+
+        Returns
+        ----------
+        total_penalty_for_single_modification: `float`
+            total penalty for this position of the dot
+        """
+        list_of_prepared_observations_and_predictions_single_fiber = []
+        for obs in range(self.n_of_obs):
+            list_of_prepared_observations_and_predictions_single_fiber_single_run = []
+            for j in range(4):
+                list_of_prepared_observations_and_predictions_single_fiber_single_run.append(
+                    self.list_of_prepared_observations_and_predictions[obs][j][i])
+
+            list_of_prepared_observations_and_predictions_single_fiber.append(
+                list_of_prepared_observations_and_predictions_single_fiber_single_run)
+
+        sigle_run_result_for_modification = []
+        try:
+            sigle_run_result_for_modification.\
+                append([i, self.quality_measure_single_spot(
+                    list_of_prepared_observations_and_predictions_single_fiber, xd, yd)])
+        except (TypeError, ValueError):
```

Comment:
Is this another case of numpy crying "wolf"? Try to catch them before you need to.